### ISSUE TITLE
Clarification of information displayed at each epoch

### ISSUE THREAD

**Issue by woctezuma** [MIC-DKFZ/nnUNet#111]
As I train the nnU-Net on some toy data, I can see the following:

```
epoch:  42
2019-12-22 18:23:23.210800: train loss : -0.9671
2019-12-22 18:23:30.078278: val loss (train=False): -0.9747
2019-12-22 18:23:30.078705: Val glob dc per class: [0.9835869986973607]
2019-12-22 18:23:31.195909: lr is now (scheduler) 0.0003
2019-12-22 18:23:31.196134: current best_val_eval_criterion_MA is 0.98350
2019-12-22 18:23:31.196233: current val_eval_criterion_MA is 0.9797
2019-12-22 18:23:31.196308: No improvement: current train MA -0.9608, best: -0.9656, eps is 0.0005
2019-12-22 18:23:31.196388: Patience: 17/50
2019-12-22 18:23:31.196459: This epoch took 148.683396 s
```

```
2019-12-22 19:41:17.579440: 
epoch:  74
2019-12-22 19:43:41.134735: train loss : -0.9746
2019-12-22 19:43:48.042461: val loss (train=False): -0.9813
2019-12-22 19:43:48.042922: Val glob dc per class: [0.9881560214781875]
2019-12-22 19:43:49.107965: lr is now (scheduler) 0.0003
2019-12-22 19:43:49.108187: current best_val_eval_criterion_MA is 0.98700
2019-12-22 19:43:49.108276: current val_eval_criterion_MA is 0.9872
2019-12-22 19:43:49.108355: saving best epoch checkpoint...
2019-12-22 19:43:49.125379: saving checkpoint...
2019-12-22 19:43:49.621252: done, saving took 0.51 seconds
2019-12-22 19:43:49.627380: New best epoch (train loss MA): -0.9719
2019-12-22 19:43:49.627527: Patience: 0/50
2019-12-22 19:43:49.627650: This epoch took 150.463133 s
```

It is not super obvious what the different variables are.

- "train loss" is the loss on the training data for the current epoch, where I assume that "the sum of the cross-entropy loss and the dice loss are used as loss function", as in your paper,
- "val loss" is the loss on the validation data,
- "Val glob dc per class" is a list of the [Dice coefficients](https://en.wikipedia.org/wiki/S%C3%B8rensen%E2%80%93Dice_coefficient#Formula) for each class. I guess "dc" stands for Dice. Moreover, I see the following formula [in your code](https://github.com/MIC-DKFZ/nnUNet/blob/bd7a1d3f25ad91e112bfe920acd29aa930473c59/nnunet/training/network_training/nnUNetTrainer.py#L577), which is akin to the formula on Wikipedia.

```python
        global_dc_per_class = [i for i in [2 * i / (2 * i + j + k) for i, j, k in
                                           zip(self.online_eval_tp, self.online_eval_fp, self.online_eval_fn)]
                               if not np.isnan(i)]
```

- "lr" is the learning rate,
- "best_val_eval_criterion_MA" is the best [exponential **m**oving **a**verage](https://github.com/MIC-DKFZ/nnUNet/blob/1f89b9f434a9f557bfa82127a98b2c7fca08dabe/nnunet/training/network_training/network_trainer.py#L510) for the evaluation criterion on the validation data, but I am not sure what the "evaluation criterion" stands for, especially since it is positive whereas the loss is negative,
- "val_eval_criterion_MA" is the moving average for the evaluation criterion on the validation data after the current epoch,
- "current train MA" is the moving average for the loss on the training data after the current epoch,
- "best" is the best moving average for the loss on on the training data,
- "eps" is the threshold for deciding that there is no improvement (if current and best are close),
- "Patience" is a counter used to control the learning rate, and which is incremented whenever there is no improvement after one epoch.

![Network training](https://i.imgur.com/25CHIEQ.png)

Could you clarify the differences between these three: i) loss vs. ii) evaluation criterion MA vs. iii) MA?

Specifically, could you explain why "evaluation criterion MA" is positive? Is "evaluation criterion MA" the moving average for the Dice score?

Finally, it seems that the best checkpoint is determined by the "val_eval_criterion_MA", and the "patience" is incremented based on the "train loss MA"? So, validation data is used to [find the best model checkpoint](https://github.com/MIC-DKFZ/nnUNet/blob/1f89b9f434a9f557bfa82127a98b2c7fca08dabe/nnunet/training/network_training/network_trainer.py#L466), and training data is used to [adapt the learning rate](https://github.com/MIC-DKFZ/nnUNet/blob/1f89b9f434a9f557bfa82127a98b2c7fca08dabe/nnunet/training/network_training/network_trainer.py#L473). But why would you use "evaluation criterion" for one, and "loss" for the other?

**Comment by woctezuma** [comment]
Meanwhile, I will dig [into the code](https://github.com/MIC-DKFZ/nnUNet/blob/1f89b9f434a9f557bfa82127a98b2c7fca08dabe/nnunet/training/network_training/network_trainer.py#L343) to try to understand.

**Comment by woctezuma** [comment]
I see that, by default, "evaluation metric" could be -1 times the loss, so that we want to minimize the loss and maximize the "evaluation metric". However, I am not sure what it is used for.

```python
        If self.all_val_eval_metrics is unused (len=0) then we fall back to using -self.all_val_losses for the MA to determine early stopping
        (not a minimization, but a maximization of a metric and therefore the - in the latter case)
```

Reference: [this line](https://github.com/MIC-DKFZ/nnUNet/blob/1f89b9f434a9f557bfa82127a98b2c7fca08dabe/nnunet/training/network_training/network_trainer.py#L422).

And these is this weird comment about the test set:

```python
            # check if the current epoch is the best one according to moving average of validation criterion. If so
            # then save 'best' model
            # Do not use this for validation. This is intended for test set prediction only.
            self.print_to_log_file("current best_val_eval_criterion_MA is %.4f0" % self.best_val_eval_criterion_MA)
            self.print_to_log_file("current val_eval_criterion_MA is %.4f" % self.val_eval_criterion_MA)
```

Reference: [this line](https://github.com/MIC-DKFZ/nnUNet/blob/1f89b9f434a9f557bfa82127a98b2c7fca08dabe/nnunet/training/network_training/network_trainer.py#L460).

**Comment by FabianIsensee** [comment]
Hi,
I see. Some more information is definitely needed here

> "train loss" is the loss on the training data for the current epoch, where I assume that "the sum of the cross-entropy loss and the dice loss are used as loss function", as in your paper,

correct

> "val loss" is the loss on the validation data,

correct

> "Val glob dc per class" is a list of the Dice coefficients for each class. I guess "dc" stands for Dice. Moreover, I see the following formula in your code, which is akin to the formula on Wikipedia.

This is the global dice, computed during validation. The TP, FP and FN are accumulated over all validation patches and the dice score is computed at the end for each class. This is an **optimistic estimate** of what the actual validation dice loss will be and I use it only to see whether the network trains at all or not. This should not be relied upon

> "lr" is the learning rate,

correct

> "best_val_eval_criterion_MA" is the best exponential moving average for the evaluation criterion on the validation data, but I am not sure what the "evaluation criterion" stands for, especially since it is positive whereas the loss is negative,

evaluation criterion here is the _Val glob dc per class_, averaged over all classes

> "val_eval_criterion_MA" is the moving average for the evaluation criterion on the validation data after the current epoch,

yes

> "current train MA" is the moving average for the loss on the training data after the current epoch,

also yes

> "best" is the best moving average for the loss on on the training data,

yup

> "eps" is the threshold for deciding that there is no improvement (if current and best are close),

yes

> "Patience" is a counter used to control the learning rate, and which is incremented whenever there is no improvement after one epoch.

yes

> Could you clarify the differences between these three: i) loss vs. ii) evaluation criterion MA vs. iii) MA?

loss is the value of the loss. That is the sum of CE and Dice loss. Best loss is -1, worst is inf.
evaluation criterion is the average of _Val glob dc per class_
MA stands for (exponential) moving average: new_smooth = alpha * old + (1 - alpha) * new

> "Do not use this for validation. This is intended for test set prediction only".

That comment is stupid and should not be there. It seems to be a leftover of some previous stuff that was there. 

> I see that, by default, "evaluation metric" could be -1 times the loss, so that we want to minimize the loss and maximize the "evaluation metric". However, I am not sure what is actually used.

The evaluation metric is always maximized. The loss minimized. If there is no evaluation metric (there always is one in nnU-Net, this is just in case I use this code for something else) then we substitute that with the -loss. 

I know this response is short - I am currently quite busy. I will update the documentation in the next year. If anything else is unclear, please let me know.

And thank you very much for your detailed issue! 

Kind regards,
Fabian

**Comment by woctezuma** [comment]
Thank you for your replies!
